kernel/generic/src/main/shutdown.c

```diff
@@ -52,5 +97,92 @@ void reboot(void)
 	halt();
 }
 
+void abort_shutdown()
+{
+	irq_spinlock_lock(&threads_lock, true);
+	thread_t *thread = atomic_load(&shutdown_thread);
+	if (thread != NULL) {
+		thread_interrupt(thread);
+		atomic_store(&shutdown_thread, NULL);
+	}
+	irq_spinlock_unlock(&threads_lock, true);
+}
+
+/* argument structure for the shutdown thread */
+typedef struct {
+	sysarg_t mode;
+	sysarg_t delay;
+} sys_shutdown_arg_t;
+
+/* function for the shutdown thread */
+static void sys_shutdown_function(void *arguments)
+{
+	sys_shutdown_arg_t *arg = (sys_shutdown_arg_t *)arguments;
+
+	if (arg->delay != 0) {
+		thread_sleep(arg->delay);
+	}
+
+	if (thread_interrupted(THREAD)) {
+		free(arguments);
+		return;
+	}
+
+	if (arg->mode == SHUTDOWN_REBOOT) {
+		reboot();
+	} else {
+		halt();
+	}
+}
+
+/* system call handler for shutdown */
+sys_errno_t sys_shutdown(sysarg_t mode, sysarg_t delay, sysarg_t kconsole)
+{
+
+#if (defined(CONFIG_DEBUG)) && (defined(CONFIG_KCONSOLE))
+	if (kconsole) {
+		grab_console();
+	}
+#endif
+
+	abort_shutdown();
+
+	/* `cancel` or default has been called */
+	if (mode != SHUTDOWN_HALT && mode != SHUTDOWN_REBOOT) {
+		return EOK;
+	}
+
+	sys_shutdown_arg_t *arg = malloc(sizeof(sys_shutdown_arg_t));
+	if (arg == NULL) {
+		return ENOMEM;
+	}
+
+	//TODO: find a better way for accessing the kernel task
```

Comment:
There is currently no other way probably, simply because there was no need for it. Moreover, global task IDs will be phased out at some point in favour of local IDs (capability handles), so this will cease to work. Not tomorrow though, but you've been warned :-)